R/crew_launcher_lsf.R

```diff
@@ -0,0 +1,287 @@
+#' @title `r lifecycle::badge('experimental')` Create a launcher with
+#'   LSF workers.
+#' @export
+#' @family launchers
+#' @description Create an `R6` object to launch and maintain
+#'   workers as LSF jobs.
+#' @details WARNING: the `crew.cluster` LSF plugin is experimental
+#'   and has not actually been tested on a LSF cluster. Please proceed
+#'   with caution and report bugs to
+#'   <https://github.com/wlandau/crew.cluster>.
+#'
+#'   To launch a LSF worker, this launcher
+#'   creates a temporary job script with a call to `crew::crew_worker()`
+#'   and submits it as an LSF job with `sbatch`. To see most of the lines
+#'   of the job script in advance, use the `script()` method of the launcher.
+#'   It has all the lines except for the job name and the
+#'   call to `crew::crew_worker()`, both of
+#'   which will be inserted at the last minute when it is time
+#'   to actually launch a worker.
+#' @inheritParams crew_launcher_cluster
+#' @param lsf_log_output Character of length 1, file pattern to control
+#'   the locations of the LSF worker log files. By default, both standard
+#'   output and standard error go to the same file.
+#'   `lsf_log_output = "crew_log_%J.log"` translates to a line of
+#'   `#BSUB-o crew_log_%J.log` in the LSF job script,
+#'   where `%J` is replaced by the job ID of the worker.
+#'   The default is `/dev/null` to omit these logs.
+#'   Set `lsf_log_output = NULL` to omit this line from the job script.
+#' @param lsf_log_error Character of length 1, file pattern for standard
+#'   error. `lsf_log_error = "crew_error_%J.err"` translates to a line of
+#'   `#BSUB-e crew_error_%J.err` in the LSF job script,
+#'   where `%J` is replaced by the job ID of the worker.
+#'   The default is `/dev/null` to omit these logs.
+#'   Set `lsf_log_error = NULL` to omit this line from the job script.
+#' @param lsf_memory_megabytes_per_cpu Positive numeric of length 1
+#'   with the megabytes of memory required.
+#'   `lsf_memory_megabytes_per_cpu = 4096`
+#'   translates to a line of `#BSUB-M 4096`
```

Comment:
From the docs at https://www.ibm.com/docs/en/spectrum-lsf/10.1.0?topic=bsub-options, it looks like `-M` is a memory limit (sounds like an upper limit) rather than memory required. I don't see a flag that sets minimum memory. If it exists, feel free to add it. Otherwise, I think we can document `-M` as an upper cap.

Also, should `#BSUB-M` be `#BSUB -M` with a space, or does it not matter?

Reply:
Argument spacing for `#BSUB` doesn't seem to matter, but for readability can include the spaces. LSF allows for minimum resource requirements to be specified using the `-R` argument (eg. `-R 'rusage[mem=16384]'`): https://www.ibm.com/docs/en/spectrum-lsf/10.1.0?topic=o-r

I'll re-document the current implementation as an upper cap. Is there a preference for a set of "core" arguments that should be explicitly included, rather than via "script_lines()"?

Reply:
Thanks! I think formal arguments may depend somewhat on the platform, but for starters, I am thinking log files, minimum memory, minimum cores, and `-cwd` (for schedulers that support it). That should be enough, but I would be open to adding more if you or other LSF users want the extra structure and documentation.